### PR TITLE
[WIP] Add wasm SIMD test

### PR DESCRIPTION
**Pull request by tlively**
Add infrastructure for testing with wasm SIMD enabled. Also add a test
that contains all SIMD instructions to see which ones can be emitted
by the toolchain and decoded by v8.

### CODE REVIEW

File: tests/test_wasm_builtin_simd.c [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+}
+i32x4 TESTFN i32x4_const(void) {
+  return (i32x4) {1, 2, 3, 4};
+}
+// v128 TESTFN v128_shuffle_interleave_bytes(v128 x, v128 y) {
+//   return __builtin_shufflevector(x, y, 0, 17, 2, 19, 4, 21, 6, 23, 8, 25, 10, 27, 12, 29, 14, 31);
+// }
+// i32x4 TESTFN v128_shuffle_reverse_i32s(i32x4 vec) {
+//   return __builtin_shufflevector(vec, vec, 3, 2, 1, 0);
+// }
```
Comment (tlively) [inline tests/test_wasm_builtin_simd.c]: Shuffles seem to crash d8.

File: tests/test_wasm_builtin_simd.c [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+}
+// i8x16 TESTFN i8x16_shl(i8x16 vec, int32_t shift) {
+//   return vec << shift;
+// }
+// i8x16 TESTFN i8x16_shr_s(i8x16 vec, int32_t shift) {
+//   return vec >> shift;
+// }
+// i8x16 TESTFN i8x16_shr_u(i8x16 vec, int32_t shift) {
+//   return (i8x16)((u8x16)vec >> shift);
+// }
```
Comment (tlively) [inline tests/test_wasm_builtin_simd.c]: Toolchain and v8 disagree about whether the shift amount should be an immediate or stack value.

File: tests/test_wasm_builtin_simd.c [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+}
+i64x2 TESTFN i64x2_neg(i64x2 vec) {
+  return -vec;
+}
+// int32_t TESTFN i64x2_any_true(i64x2 vec) {
+//   return __builtin_wasm_any_true_i64x2(vec);
+// }
+// int32_t TESTFN i64x2_all_true(i64x2 vec) {
+//   return __builtin_wasm_all_true_i64x2(vec);
+// }
```
Comment (tlively) [inline tests/test_wasm_builtin_simd.c]: LLVM doesn't know how to split the operands of these ops up when i64x2 ops are disabled.

File: tests/test_wasm_builtin_simd.c [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+}
+i32x4 TESTFN i32x4_trunc_u_f32x4_sat(f32x4 vec) {
+  return __builtin_wasm_trunc_saturate_u_i32x4_f32x4(vec);
+}
+// i64x2 TESTFN i64x2_trunc_s_f64x2_sat(f64x2 vec) {
+//   return __builtin_wasm_trunc_saturate_s_i64x2_f64x2(vec);
+// }
+// i64x2 TESTFN i64x2_trunc_u_f64x2_sat(f64x2 vec) {
+//   return __builtin_wasm_trunc_saturate_u_i64x2_f64x2(vec);
+// }
```
Comment (tlively) [inline tests/test_wasm_builtin_simd.c]: LLVM doesn't know how to split these either.

**Comment by tlively** [comment]
This test additionally passes `binaryen0` and `wasmobj0` modes with the LLVM patch up for review here: https://reviews.llvm.org/D55956

File: tests/test_wasm_builtin_simd.c (hunk 1 above)
Comment (dtig) [reply tests/test_wasm_builtin_simd.c]: I'll take a look to see what's happening.

File: tests/test_wasm_builtin_simd.c (hunk 2 above)
Comment (dtig) [reply tests/test_wasm_builtin_simd.c]: Will prototype as stack values in V8, and report back on the open issue.

File: tests/test_wasm_builtin_simd.c [hunk 5]
```diff
@@ -0,0 +1,862 @@
...
+
+  // expected {0x00ffffff, 0xff00ffff, 0xff00ffff, 0xff00ffff},
+  //      got {0x0000ffff, 0xffffffff, 0x0000ffff, 0xffffffff}
+  expect_vec(
+    i8x16_le_u(
+      (i8x16){0, 127, 13, 128,  1,  13, 129,  42, 0, 127, 255, 42,   1,  13, 129,  42},
+      (i8x16){0, 255, 13, 42, 129, 127,   0, 128, 0, 255,  13, 42, 129, 127,   0, 128}
+    ),
+    ((i8x16){-1, -1, -1, 0, -1, -1, 0, -1, -1, -1, 0, -1, -1, -1, 0, -1})
+  );
```
Comment (tlively) [inline tests/test_wasm_builtin_simd.c]: V8 is giving incorrect results here.

File: tests/test_wasm_builtin_simd.c [hunk 6]
```diff
@@ -0,0 +1,862 @@
...
+  // expected {0x00ffffff, 0xff00ffff, 0xff00ffff, 0xff00ffff},
+  //      got {0x0000ffff, 0xffffffff, 0x0000ffff, 0xffffffff}
+  expect_vec(
+    i8x16_ge_u(
+      (i8x16){0, 127, 13, 128,  1,  13, 129,  42, 0, 127, 255, 42,   1,  13, 129,  42},
+      (i8x16){0, 255, 13, 42, 129, 127,   0, 128, 0, 255,  13, 42, 129, 127,   0, 128}
+    ),
+    ((i8x16){-1, 0, -1, -1, 0, 0, -1, 0, -1, 0, -1, -1, 0, 0, -1, 0})
+  );
```
Comment (tlively) [inline tests/test_wasm_builtin_simd.c]: V8 is giving incorrect results here, too.

File: tests/test_wasm_builtin_simd.c (hunk 5 above)
Comment (tlively) [reply tests/test_wasm_builtin_simd.c]: i8x16.le_u incorrectly returns 0 in lane comparing 13 with 13, -1 in lane comparing 129 with 0, and 0 in lane comparing 42 with 128.

Comment (tlively) [reply tests/test_wasm_builtin_simd.c]: I find it strange that the lane comparing 0 with 0 is fine but the lane comparing 13 with 13 is wrong :/

File: tests/test_wasm_builtin_simd.c (hunk 6 above)
Comment (tlively) [reply tests/test_wasm_builtin_simd.c]: The source comment here is wrong due to a copy paste error. The actual results here are that i8x16.ge_u incorrectly returns 0 in the lanes comparing 0 to 0, and 0 in the lanes comparing 129 to 0.

**Comment by gabrielcuvillier** [comment]
Out of curiosity, how do you enable SIMD in wasm? When I try to do this, the compiler says it is only working with asm.js

File: tests/test_wasm_builtin_simd.c (hunk 5 above)
Comment (dtig) [reply tests/test_wasm_builtin_simd.c]: Took a quick look, and we already comparison tests that test these values, so this is weird to me too. Could you file a V8/Chromium for tracking with a repro?

**Comment by tlively** [comment]
@gabrielcuvillier you need tip of tree LLVM and Binaryen, but otherwise it should just work with the wasm backend and `-s SIMD=1`. What specific error are you getting?

File: tests/test_wasm_builtin_simd.c (hunk 5 above)
Comment (tlively) [reply tests/test_wasm_builtin_simd.c]: https://bugs.chromium.org/p/v8/issues/detail?id=8635

**Comment by tlively** [comment]
@kripken How would you feel about landing this with the broken tests commented out to let others play around with it?

**Comment by kripken** [comment]
sgtm, sure.

Is the plan to eventually remove the ASM_SIMD as we get those tests passing in wasm? Or is there a reason we'd need that distinction in the long term too?

**Comment by gabrielcuvillier** [comment]
When I use -msse2 option with the compiler (for SSE2 intrinsics), I have the error "SIMD is used, but not supported in WASM mode yet" 

Should I also use the SIMD=1 flag too ? I don't particularly want autovectorization, I a more interested in compiling existing SSE/SSE2 intrinsics code

**Comment by tlively** [comment]
@gabrielcuvillier Oh gotcha. The SIMD support being tested here is more basic than that, so SSE intrinsics and other non-wasm SIMD intrinsics are not expected to work yet. We may implement a compatibility layer to allow other platforms' intrinsics to work with wasm in the future, but we're still working on getting the basic codegen and execution of wasm SIMD instructions working for now.

@kripken Maybe the distinction will be removed in the future? The only similarity between the ASM_SIMD and WASM_SIMD tests for now is that they both use `-s SIMD=1`, so a lot of work will have to be done before that can happen.

**Comment by tlively** [comment]
This should pass once https://reviews.llvm.org/D56520 lands.

**Pull request merged**